### PR TITLE
Annotate metrics for /service/... upstream location

### PR DESCRIPTION
**Pull request by ahairape**
This PR provides annotation for metrics for services accessible under /service/... location via Admin Router.

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

https://jira.mesosphere.com/browse/DCOS_OSS-4950

## Checklist for all PRs

  - [ ] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change:
  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

### CODE REVIEW

**Comment by ahairape** [comment]
@mesosphere-mergebot label Ready For Review

File: packages/adminrouter/extra/src/includes/metrics-service-location-level.conf [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+# Create a counter for every URI that requests go to when matching the location directive that this filter appears in.
+# Tag the counter with the manually set service name, the backend that the request was passed on to and the resulting status code.
+vhost_traffic_status_filter_by_set_key uri:=$original_uri service:=${service_tag}_._._backend:=${upstream_addr}_._._status:=${status};
+
+# Create a counter for every user agent that requests were sent from which match the location directive that this filter appears in.
+# Tag the counter with the manually set service name, the backend that the request was passed on to and the resulting status code.
+vhost_traffic_status_filter_by_set_key useragent:=$http_user_agent service:=${service_tag}_._._backend:=${upstream_addr}_._._status:=${status};
+
+# Another thing to keep in mind is the consideration of whether to import this file before
```
Comment (timaa2k) [inline packages/adminrouter/extra/src/includes/metrics-service-location-level.conf]: This section is no longer true and should be removed.

File: packages/adminrouter/extra/src/includes/metrics-service-location-level.conf [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# identify the upstream that the metrics belong to. This constitutes a mechanism to
+# annotate the Admin Router config, allowing one to collect metrics of dynamically
+# resolved upstreams on the fly. It is also the name that will show up in Grafana when
+# referring to a particular upstream. Grouping metrics based on upstream would normally
+# only be possible with Nginx Plus because only then they can be defined in a upstream
+# directive and continue to be dynamically resolved.
+# https://tenzer.dk/nginx-with-dynamic-upstreams/
+#
+# Before including this file, Nginx variables $upstream_tag and $original_uri must be set like this:
+# set $upstream_tag <CUSTOM_UPSTREAM_TAG>;
```
Comment (timaa2k) [inline packages/adminrouter/extra/src/includes/metrics-service-location-level.conf]: `s/$upstream_tag/$service_tag/`
`s/CUSTOM_UPSTREAM_TAG/CUSTOM_SERVICE_TAG/`

File: packages/adminrouter/extra/src/lib/service.lua [hunk 3]
```diff
@@ -313,6 +313,10 @@ local function recursive_resolve(auth, path, marathon_cache, mesos_cache)
     local err_code = nil
     local err_text = nil
 
+    -- The variable service_name must be declared local here,
+    -- see https://jira.mesosphere.com/browse/DCOS_OSS-4999 for details
```
Comment (jongiddy) [inline packages/adminrouter/extra/src/lib/service.lua]: I'm not sure this needs a comment. It's just how it should have been done.

File: packages/adminrouter/extra/src/test-harness/tests/open/test_metrics.py [hunk 4]
```diff
@@ -28,14 +30,17 @@ class TestMetrics:
         ids=list(get_static_upstream_annotations().values())
         )
     def test_metrics_prometheus_static_upstreams_annotated(
-            self, master_ar_process, location, annotation):
+            self, master_ar_process, jwt_generator, location, annotation):
```
Comment (jongiddy) [inline packages/adminrouter/extra/src/test-harness/tests/open/test_metrics.py]: Can we use the `valid_user_header` fixture here to keep things simpler?

**Comment by timaa2k** [comment]
The ACCEPTANCE CRITERIA also mention an integration test that `nginx_service_*` metrics are reported through Telegraf. Given the SOAK update I think we can do this in a follow-up PR after it has been confirmed manually.

**Comment by jgehrcke** [comment]
@ahairape why did you ignore the checkbox "Added a comprehensible changelog entry to CHANGES.md or explain why this is not a user-facing change:"?

**Comment by jgehrcke** [comment]
> @ahairape why did you ignore the checkbox "Added a comprehensible changelog entry to CHANGES.md or explain why this is not a user-facing change:"?

Valid answer: this patch here will be released with 1.13.0, which will be released after 1.12.4, so that the relevant entry only needs to go into the 1.12.4 CHANGES.md.

Correspondingly, I requested a change to https://github.com/dcos/dcos/pull/4990.

**Pull request merged**